source/common/SyclGraph.cpp

```diff
@@ -0,0 +1,92 @@
+/** @file   SyclGraph.cpp
+ *  @author Yantao Zhang (yantao.zhang@intel.com)
+ *  @brief  Implementation of SYCL graph 
+ *  @todo   Update the executable graph if needed.
+ */
+
+#include "SyclGraph.h"
+
+namespace tinydpcppnn{
+
+inline std::deque<SyclGraph*>& current_captures() {

```

Comment:
I am not sure these are necessary. If I am following the implementation below correctly, then you cannot nest the scope guards. I.e.:
if the stream is already recording, an empty scope guard is returned and no new recording is started.
Thus, the case where current_captures().back() != this can never occur. Even more, currently no more than 1 SyclGraph per queue is possible with the current implementation due to this.
Am I missing something?